### ISSUE TITLE
Python Script plugin is not compatible with Notepad++ version 7.6

### ISSUE THREAD

**Issue by dnz3d4c** [bruderstein/PythonScript#109]
- Python Script plugin is not compatible with Notepad++ version 7.6
  1. If you create a plugin folder as instructed at <https://notepad-plus-plus.org/news/notepad-7.6-released.html>, a dialog box will appear stating that the plugin is incompatible.

Thanks.

**Comment by chcg** [comment]
@dnz3d4c Where did you place the python27.dll? It is still required to be at the level of the n++.exe.
So python script plugin up to current 1.3 is not fully compatible with N++ 7.6.

**Comment by ClaudiaFrank** [comment]
@dnz3d4c
Seems to work if you do it that way.
In general, however, as chcg already said, certain stuff would have to be adapted to make it
fully compatible with this new plugin structure.

![grafik](https://user-images.githubusercontent.com/16170444/48666557-c6869b00-eac4-11e8-9901-2e041e2ee9c3.png)

**Comment by YeltsinReyes** [comment]
@dnz3d4c How do I work for the plugin Emmet and PythonScript?

**Comment by bitagoras** [comment]
I developed a Notepad++ plugin called PyPadPlusPlus, which is based on PythonScript. Since the installation is very cumbersome, especially now with Notepad++ v7.6, I published it in a bundle with Npp7.6 and PythonScript1.3.0.0 as a portable version. So you don't need the AppData folder at all. It also includes the new plugin manager. Thank to your comment @ClaudiaFrank I finally found out how to do it. Thank you a lot!
You can download this special portable edition of Notepad++ v7.6 with PythonScript and PyPadPlusPlus pre-installed here:
https://github.com/bitagoras/PyPadPlusPlus

**Comment by diamondpointmedia** [comment]
Please fix asap. I recently upgraded to 7.6.3 only to find out that my procedure stopped working. After screwing around for a couple of hours and still not being able to get it to work I had to revert back to 7.5.9.

**Comment by vsrawat** [comment]
There is nothing wrong due to pythoncript.

Notepad++ has made some changes, then reverted, and made other changes, 
that has lead to this problem.

The changes in Notepad++ are still not final, they are going to make at 
least one more change in which this problem might occur, if you adopt to 
the current situation.

It is suggested that you do revert to 5.9 and use that till the dust 
settles and others confirm that there is no problem with npp++ and 
pythonscript combo.

If you do want to use the latest releases of both, I suggest you go to 
npp forum and ask there, you will get more timely detailed information 
there.

Thanks.
-- 
Rawat

**Comment by Rubiya99** [comment]
@vsrawat 
Thank you for the info,
I have been trying to install emmet and python script in the current version (7.6.4) for the past few days and it was not installing , i thought there was a problem with my laptop but after reading this forum i have the 7.5.9 version downloaded in which emmet and pythonscript are working perfectly

**Comment by vsrawat** [comment]
You are welcome.

I have installed 7.6.2 and adjusted to their new settings for plugins, 
and then they released 7.6.3 in which they changed settings again, so I 
got fed up and didn't install 7.6.3 at all. My work suffers if something 
goes wrong in npp.

I am in npp forum, so when they would release a next version in which 
everything settles down and is permanent, then I would install that 
version, and then I would post here also, and then you can install that. 
I will also post what changes are to be done.

Thanks.
-- 
Rawat

**Comment by SagePtr** [comment]
Same problem, Notepad++ 7.6.3 simply doesn't see PythonScript and don't show it in plugins menu, neither Emmet (depends on PythonScript) do. Reverting to Notepad++ 7.5.9 solved issue, but i don't want to stick to outdated Notepad++ version.

**Comment by vsrawat** [comment]
Notepad++ had made changes to its arrangement of all plugins.

I think now you have make a folder of the plugin in and put the plugin's 
  dll in that.

Still, as python script has two dll, one dll still remains in 
Notepad++'s Program files folder.

I am also waiting for them to lock the position of all files, then I 
will update, but they have recently released 7.6.4 in which they have 
not made any changes in that fluid arrangement nor have said anything at 
all, seems they have left it in intermediate fluid location.

ok. I will update to 7.6.4, and see which file has to go where, and then 
shall post it here by tomorrow.

Then you all can update. Please wait for a day. I am on 64 bit w10.

Thanks.
-- 
Rawat

**Comment by vsrawat** [comment]
I have updated to w-10-64 bit npp 7.6.4

The only change from previous npp 7.5.9 or earlier will be

in the Program Files\Notepad++\plugins folder,

previously, all plugins dll were residing just there,
now, under plugin folder, you have to create a new folder with the name 
of the plugin (removing .dll) and move the plugin.dll to that folder.

so, for pythonscript
1. go to C:\Program Files\Notepad++\plugins\
2. create a folder named PythonScript
3. move PythonScript.dll to this folder

Do the same for all your plugins one by one.

NOTE: python27.dll still remains in C:\Program Files\Notepad++
don't more that to newly created folder.

I think this part should be taken care by pythonscript developers that 
they update pythonscript so that all py files are residing only in 
C:\Program Files\Notepad++\plugins\PythonScript folder.

Most of the plugins will have only single dll file, but only a few, like 
  pythonscript have multiple files/ folder in that newly create folders.

my own PythonScript also has lib and scripts folders, I am not sure 
whether they were moved by me, or were installed by py installer.

so, After you create the above folder and move dll to the said folder, 
it is better if you run pytonscript installer again, or run npp and 
install pythonscript afresh using plugin admin, that would put the 
missing files if any, or will put duplicate that will not harm.

Do write if any problems comes after that, we will analyze and suggest 
solutions.

Thanks.
-- 
Rawat

**Comment by sheikhmishar** [comment]
Good news, guys. I've managed to use latest emmet, pythonScript 1.3 and pypadplusplus on notepad++ 7.6.2. So excited to share with you guys. Just give me some moment to test and then upload

**Comment by sheikhmishar** [comment]
**Here you can check:** [https://github.com/sheikhmishar/notepad-plus-plus-mod ](https://github.com/sheikhmishar/notepad-plus-plus-mod )

**Comment by MetaChuh** [comment]
in case anyone needs the step by step guide on what to download and how to install pythonscript on notepad 7.6.3, 7.6.4 and above, here's the latest copy of the ...
[Guide: How to install the PythonScript plugin on Notepad++ 7.6.3, 7.6.4 and above:](https://notepad-plus-plus.org/community/topic/17256/guide-how-to-install-the-pythonscript-plugin-on-notepad-7-6-3-7-6-4-and-above)
from the notepad++ community forum:

#### complete guide how to install pythonscript on notepad++ 7.6.3, 7.6.4 and above:

-----

**for notepad++ 7.6.4, 64 bit (installed version):**

* download and extract PythonScript_Full_1.3.0.0_x64.zip from [>>> here <<<](https://github.com/bruderstein/PythonScript/releases/download/v1.3.0/PythonScript_Full_1.3.0.0_x64.zip) to your desktop.
note: do not use any other available release type, except this zip.

* open `%ProgramFiles%\Notepad++\plugins\` in windows explorer and create a folder called `PythonScript`.

* copy **PythonScript.dll** from the plugins folder of this extracted zip to:
`%ProgramFiles%\Notepad++\plugins\PythonScript\PythonScript.dll`

* copy **python27.dll** from this extracted zip to:
`%ProgramFiles%\Notepad++\python27.dll`

* copy the folders `scripts` containing machine level scripts and `lib` containing python libraries, from the zip's plugins\PythonScript folder to:
`%ProgramFiles%\Notepad++\plugins\PythonScript\`

-----

**for notepad++ 7.6.4, 32 bit (installed version):**

* download and extract PythonScript_Full_1.3.0.0.zip from [>>> here <<<](https://github.com/bruderstein/PythonScript/releases/download/v1.3.0/PythonScript_Full_1.3.0.0.zip) to your desktop.
note: do not use any other available release type, except this zip.

* open `%ProgramFiles(x86)%\Notepad++\plugins\` in windows explorer and create a folder called `PythonScript`.

* copy **PythonScript.dll** from the plugins folder of this extracted zip to:
`%ProgramFiles(x86)%\Notepad++\plugins\PythonScript\PythonScript.dll`

* copy **python27.dll** from this extracted zip to:
`%ProgramFiles(x86)%\Notepad++\python27.dll`

* copy the folders `scripts` containing machine level scripts and `lib` containing python libraries, from the zip's plugins\PythonScript folder to:
`%ProgramFiles(x86)%\Notepad++\plugins\PythonScript\`

-----

**additional note:** if you have installed notepad++ to a different location than the default, please adapt `%ProgramFiles(x86)%\Notepad++`or `%ProgramFiles%\Notepad++` to your custom path.

**Comment by Gitoffthelawn** [comment]
> download and extract PythonScript_Full_1.3.0.0.zip from >>> here <<< to your desktop.
note: do not use any other available release type, except this zip.

Just FYI, the PythonScript_Min_1.3.0.0.zip appears to work equally as well.

**Comment by MetaChuh** [comment]
@Gitoffthelawn 

> Just FYI, the PythonScript_Min_1.3.0.0.zip appears to work equally as well.

yes, you are absolutely correct.

the notepad++ community guide has been evolving through various forum user requests, where some required the presence of the samples.

the full `PythonScript_Full_1.3.0.0.zip` release was chosen as "the one" to install, to cover all past community needs with one single guide, e.g. to avoid repeated questions about missing pythonscript samples/templates.

i apologise if this approach might have created any confusion.

**Comment by RomanStone** [comment]
Notepad 7.6.6 just updated (17.04.2019)

copied dlls to
```
%ProgramFiles(x86)%\Notepad++\plugins\PythonScript\python27.dll
%ProgramFiles(x86)%\Notepad++\plugins\PythonScript\PythonScript.dll
```

other instructions not work for me (Win7 x64)

**Comment by Gitoffthelawn** [comment]
@RomanStone Move `python27.dll` to `%ProgramFiles(x86)%\Notepad++\python27.dll`.  That should work.

**Comment by RomanStone** [comment]
Yes. But without `PythonScript.dll` file copy in `%ProgramFiles(x86)%\Notepad++\plugins\PythonScript\PythonScript.dll` 
`Python Script` plugin not listed in `Plugins` sub menu. 

hope this will help someone

Win7 x64 SP1 (w/latest updates, VC 4.1), NPP  7.6.6 (x86), Plugin Manager 1.4.12.0, Python Script 1.3.0.0 (Installed via Plugin Manager)

**Comment by MetaChuh** [comment]
many thanks for sharing your experience, @RomanStone 

> NPP 7.6.6 (x86), Plugin Manager 1.4.12.0, Python Script 1.3.0.0 (Installed via Plugin Manager)

please note that plugin manager is not compatible with the plugin folder structure of notepad++ versions above 7.5.9.

both plugin installers serve their purpose at the respective versions.
the best practice is, to use `plugin manager` for versions 7.5.9 and below, and the built in `plugins admin` available at versions 7.6.3 and above (plus manual installs for compatible plugins, like pythonscript, that are not yet listed).

best regards.

**Comment by Gitoffthelawn** [comment]
@MetaChuh wrote "...plus manual installs for compatible plugins, like pythonscript, that are not yet listed.."

Do you know why it isn't listed yet?

**Comment by MetaChuh** [comment]
@Gitoffthelawn 

i have read, that one of the reasons is, that python[xx].dll currently has to be deployed to the notepad++ root folder, which contradicts the one folder plugin deployment design.

the latest things i know are from here:
https://notepad-plus-plus.org/community/topic/17430/pythonscript-2-0/5

**Comment by Gitoffthelawn** [comment]
Thanks @MetaChuh.  It's odd that python[xx].dll need to be deployed there.  There must be some big challenge involved in just deploying it with the rest of the plugin.  Do you or @chcg have any info as to why it is so challenging?

**Comment by MetaChuh** [comment]
@Gitoffthelawn 

many thanks for your questions, i hope i can answer some of them.

the parent process and main executable is notepad++.exe.
python.dll has to be at that same location, or alternatively at a location within the system's path.
same applies to e.g. applications requiring msvcr100.dll, or other runtime dlls, which have to be at the same location as the parent main executable, or at the system's path, as a registered library.

possibly helpful answer, to why can all plugin dlls be loaded from somewhere else, but python.dll not:
the plugins, which do reside at other locations as the main executable, get explicitly loaded from their different locations by notepad++.exe at startup.
but if any of them would require any other runtime library, this library would also have to be present at either the same location as the main executable, or at the system's path, as a registered library.

as far as i've been told, it would require a major rewrite (if possible at all) to use python.dll outside the path of the main notepad++ executable, with the only other hardcore option to build pythonscript.dll with the complete python library statically embedded.

deploying (copying) python.dll to the path of the main executable notepad++.exe, by the pythonscript plugin itself, after install, would currently be the only quick (but dirty) fix.
this is what users are currently doing manually, as this is basically the only thing that is needed to get pythonscript running in newer versions of notepad++, e.g. version 7.6.6.

best regards

**Comment by Gitoffthelawn** [comment]
@MetaChuh Thanks!  That answers my question perfectly!

**Comment by vsrawat** [comment]
long winded answer.

To me, this reason seems more down to earth, that (methink) all other 
plugins have single dll, so those were moved to new locations, but 
pythonscript had two dlls, so, the persons working on this shifting just 
missed this dll, and moved the other dll to right location.

When all other dlls can be loaded from new location, why can't just one 
dll of python script can't be?

This brings to me to question: When all other plugins can work with a 
single dll, why do we need two for py? Can't both merged into a single dll?

Thanks.
-- 
Rawat

**Comment by sasumner** [comment]
@vsrawat You have demonstrated that you have no clue.  Please just accept the stated reasons.

**Comment by vsrawat** [comment]
I am stupid but your great knowledge doesn't change the fact that users 
of npp as well as py are suffering from years due to wrong arrangement 
of software and you couldn't improve it.

Seems no one is the npp or py developers' group cares for endusers. Now 
both the software have become self-pride of developers and end users 
have become unavoidable burden.

People like you have thinking shallow to get hurt by small comment 
without realizing that such remarks are intentionally made to shake you 
out of your comfort zone so that you do something to improve the 
situation. But no, you would just retort and close the chapter, you 
can't contribute anything anywhere to reduce entropy.

Thanks.
-- 
Rawat

**Comment by MetaChuh** [comment]
@vsrawat 

you are not stupid at all.
your input has often helped us to rethink procedures and adapt them to the needs of others.

most important: you share knowledge you adopt, instead of just using it for yourself.
this is much appreciated and very welcome.

many thanks and best regards.

ps: sorry to all for this off topic post.

**Comment by sasumner** [comment]
@vsrawat It is fine to be a user and decree "this should be thus" and "that should be this way", but when there are current technical reasons why things cannot be made better at the moment, it doesn't help that users that aren't going to contribute to the fixes make decrees.  You're not stupid; but it seemed that you just disregarded the technical limitations that are in effect at the moment.

**Comment by chcg** [comment]
Probably something in:

https://github.com/bruderstein/PythonScript/blob/master/PythonScript/src/PythonHandler.cpp#L84

needs to be modified.

Static linkage is not used to be able to use an own version of python instead of the one bundled with pythonscript.

**Comment by ClaudiaFrank** [comment]
@chcg - I don't think so. Recently I was working with some customer of us which had a similar issue and
if I'm right then the only possible way to achieve that pythonXX.dll gets loaded from a different dir is to 
change npps [LoadLibrary](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/1dfa92c4b7e3f4acb26fe6dbfb632618c7a21d43/PowerEditor/src/MISC/PluginsManager/PluginsManager.cpp#L133) behavior by using LoadLibraryEx with [LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR](https://docs.microsoft.com/en-us/windows/desktop/api/libloaderapi/nf-libloaderapi-loadlibraryexa), then pythonXX.dll and pythonscript.dll should be in the same directory.

**Comment by Ekopalypse** [comment]
@ClaudiaFrank - I can confirm that using LoadLibraryEx with this flag works but only if WindowsXP support
is dropped, meaning a build with WindowsXP compatibility is not possible anymore.
And, by the way, let me say thank you for all of the work you've done in the past. I really learned a lot from your examples.

**Comment by ClaudiaFrank** [comment]
@Ekopalypse, WindowsXP support, really :-)  Windows 7 is already dying :-D
Thank you for your kind words

**Comment by sasumner** [comment]
We need to get Pythonscript and, more importantly, Notepad++ itself to drop XP support...as in it should be made to not even open on XP!  I thought it was dead a few times but unfortunately it keeps coming back.

Side note: Regarding Windows 7 dying, my employer (one of the biggest in the world) will not switch from Windows 7 until there is a viable alternative, and they have come out and said that Windows 10 (or even 8) are NOT viable alternatives.

**Comment by Ekopalypse** [comment]
Just to make clear, I'm thinking the same - get rid of XP. 
I was mentioning it only as it seems to be the default compiler setting.

**Comment by chcg** [comment]
@ClaudiaFrank You are right the python dll is already loaded before DllMain() is called, so:
- N++ needs to be modified (hard to achieve, I guess)
- the loading\linking to python dll needs to be modified 
- python lib needs to be statically linked in

**Comment by ClaudiaFrank** [comment]
@chcg, but statically linking ... you know what that means ... no more existing 3rd party dlls, no more local python installs ... no more fun :-D

If Don can't be convinced, I assume runtime linking must be used which, I guess, means splitting npp plugin and python related stuff in its own dlls. Well I guess that would be a good chance to jump to python3 :-D  ( I know, easier said than done )

**Comment by bitagoras** [comment]
In the release notes of the new Notepad++ v7.7 it says:
* Upgrade Scintilla from v3.56 to v4.14.

Has this some impact on PythonScript? I tested it and everything seems to work with my scripts.

My setup: I installed Notepad++ v7.7 with the option not to use "AppData". The installer of PythonScript v1.4.0.0 still places the `PythonScript.dll` directly in the `plugins/` folder. You have to move it manually to the subfolder `plugins/PythonScript/` to make it work.

**Comment by Ekopalypse** [comment]
It has, see [here](https://notepad-plus-plus.org/community/topic/17613/notepad-7-7-release-candidate/6) and [here](https://notepad-plus-plus.org/community/topic/17622/notepad-7-7-release-candidate-v3/2) for more infos.

**Comment by chcg** [comment]
@ClaudiaFrank With python 3 the same dll dir issue arises, so this step will not be easier with switching.
So the next step will be to regenerate the stuff with the new scintilla iface file of 4.14. To figure out what is broken now and needs to be adapted.

**Comment by Gitoffthelawn** [comment]
Should people who are using the pythonscript plugin upgrade to Notepad++ version 7.7, or hold off?

**Comment by sasumner** [comment]
@Gitoffthelawn I say "go for the 7.7 upgrade!" :)

**Comment by Ekopalypse** [comment]
@sasumner, but doesn't that mean you loose the editor callbacks, if you don't want to click the error dialog everytime it gets called?

**Comment by sasumner** [comment]
@Ekopalypse Not sure what you mean...  I'm using 7.7 with PS1.3 seemingly very successfully (PS1.4 change list made me say "meh, when I get around to it").  No clicking of error dialogs, no losing of callbacks...

**Comment by Ekopalypse** [comment]
Strange, when I'm using npp7.7 and PS I always get a popup when a callback is called.
Like this 
![image](https://user-images.githubusercontent.com/47723516/58093372-e86cf180-7bbd-11e9-91e3-bdd94774f0e8.png)
which, from my very little c++ understanding makes sense as the new [scintilla notification structure](https://www.scintilla.org/ScintillaDoc.html#Notifications) contains an additional field **int listCompletionMethod** now

**Comment by Ekopalypse** [comment]
I 've downloaded [npp_7.7.0](https://notepad-plus-plus.org/repository/7.x/7.7/npp.7.7.bin.x64.7z) and extracted it to a new directory. Then I downloaded [PS_1.3.0](https://github.com/bruderstein/PythonScript/releases/download/v1.3.0/PythonScript_Full_1.3.0.0_x64.zip) and unzipped it in the main npp.exe directory.
Moved PythonScript.dll into plugins\PythonScript directory and started npp. So far so good.
Open console -> no error.
Executed the following two lines in the console

`from __future__ import print_function
editor.callback(lambda a: print(a), [SCINTILLANOTIFICATION.MODIFIED])`

console returned **True**. 
Now as soon as I try to type some text I receive the error.
What am I missing?

**Comment by sasumner** [comment]
@Ekopalypse said
>What am I missing?

I'm not sure.  I have a startup.py that installs a lot of stuff that remains resident...synchronous callbacks, asynchronous callbacks...  All...that I notice on a continual basis...is working fine.  I was in fact totally surprised by this.  Occasionally I check the `?` menu to make sure I am really running 7.7.  ;)

I just copied my PS1.3 stuff over from my portable 7.6.6 into the directory structure for my portable 7.7, but it sounds like you did something very similar.  One difference is I'm using 32-bit; you may want to try that as I see your links above are 64-bit?

**Comment by Ekopalypse** [comment]
You are correct, using the 32bit version is working without error!! 
Phew, what is going on here?? How can PS receive a notification structure larger than expected?
Hmmm ... I would argue there is a problem in 32bit code ... or MORE LIKELY - I still have no clue about C++

**Comment by ClaudiaFrank** [comment]
@Ekopalypse - it might be hard to begin with C++ but I would say it is worth doing it as you get more
detailed information how most of the things work under the hood.

**Comment by Ekopalypse** [comment]
@chcg - thank you for clarification - this makes sense, I guess. :-)
@ClaudiaFrank - ... but C++ is so strange compared to python but I see what you mean.

**Comment by ClaudiaFrank** [comment]
@Ekopalypse - why not using PythonScript plugin project to get started.
I'm sure @chcg would love to get some help migrating to the new scintilla/npp version.
It's not that difficult, and as you seem to have the python knowledge how to use scintilla it might be even easier. 
That's how I started building c++ knowledge - still a lot to do but ... 
Building knowledge isn't wasting time.

**Comment by Ekopalypse** [comment]
@chcg - [this](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/5802) would solve the issue,  with python.dll have to be in notepad++ directory, right?

**Comment by chcg** [comment]
Waiting on integration of https://github.com/notepad-plus-plus/notepad-plus-plus/pull/5853 to N++.

**Comment by chcg** [comment]
See #129, created https://github.com/bruderstein/PythonScript/releases/tag/v1.5.0 as intermediate version containing the interface changes. The changed packaging to reflect also the changes in N++ folder structure is available with PR #131

**Comment by chcg** [comment]
See https://github.com/bruderstein/PythonScript/releases/tag/v1.5.1

**Comment by Ekopalypse** [comment]
Downloaded 1.5.1 full zip and extracted to ...\plugins\PythonScript
Two questions if you don't mind:
1.) docs folder goes into the binary folder, where the dlls are, shouldn't it be in the config folder?
Personally I don't care, just wanted to make sure it was intended
2.) I run the tests from the repo an got
Tests Run: 259 Errors: 2 Failures: 15
If you don't mind I can have a look at it and maybe include missing tests.

If there is anything else I can do, let me know.

**Comment by chcg** [comment]
@Ekopalypse Your help is highly welcome.
Ad 1, With PluginAdmin the zip content could just be extracted into the plugin dir. So config folder is no option for installation.
Ad 2, Result seem to match with my numbers. The main part is working and most of the failing tests were already failing with N++ 7.5.9 + PS 1.4.0. So unittests are still work in progress to get them running error free. I expect that there are still some hidden issues with int (32bit) vs. intptr_t(64bit) in the updated ScintillaWrapper class for the x64 plugin version.

Further open points:
- test sample scripts, if they are still working
- check docu for missing updates (currently I update the package structure image)

@sasumner You might want to have a look, if the new release 1.5.1 is also working for you.

**Comment by Ekopalypse** [comment]
@chcg - current status Tests Run: 259 Errors: 0 Failures: 0
I'm creating a PR next week.
A few samples need some changes here and there - not a big issue - I can do this next week as well.

What do you mean by checking docu for missing updates? Which updates? PS updates?

**Comment by chcg** [comment]
@Ekopalypse Thanks for your upcoming PR and work.
Sorry with updates I had in mind corrections, removing/updating outdated informations. Already did a checkin https://github.com/bruderstein/PythonScript/commit/dd55268d77e8a884bfaf14b636ca1134885dd700 for that.

Another point to work on:
- probably https://github.com/bruderstein/PythonScript/blob/master/PythonScript/src/CreateWrapper.py could be extended to also create at least some templates for the python tests

**Comment by Ekopalypse** [comment]
Will have a look, but from my experience I don't think you can gain a lot of benefits as most functions
need to interact with other functions to produce something useful.

**Comment by Ekopalypse** [comment]
@chcg - the following scintilla methods do have a different function layout 

1.4.0
braceMatch( (Editor)arg1, (int)pos) -> int :
createDocument( (Editor)arg1) -> int :
createLoader( (Editor)arg1, (int)bytes) -> int :
getPropertyInt( (Editor)arg1, (object)key) -> int :

1.5.1
braceMatch( (Editor)arg1, (long)pos, (int)maxReStyle) -> int :
createDocument( (Editor)arg1, (long)bytes, (int)documentOptions) -> int :
createLoader( (Editor)arg1, (long)bytes, (int)documentOptions) -> int :
getPropertyInt( (Editor)arg1, (object)key, (int)defaultValue) -> int :

That's the reason why some of the samples didn't work anymore and might break existing code.

Currently I see 3 ways to go:
1. keep it like it is, new scintilla not backward compatible anyhow and inform in docu and on download page about those changes.
2. create the wrapper function with a default value for the new parameters
3. create a overridden wrapper function and mark the old one deprecated

**Comment by chcg** [comment]
@Ekopalypse 
I think 1. is ok. Otherwise educated guesses need to be made.

Which is probably ok for:
> 
> # Find the position of a matching brace or INVALID_POSITION if no match.
> # The maxReStyle must be 0 for now. It may be defined in a future release.
> fun position BraceMatch=2353(position pos, int maxReStyle)

, but how should be achieved that this TODO is not forgotten in the future.

**Comment by Ekopalypse** [comment]
@chcg - sorry for asking this but what is exactly the way to open PR?
What I did:
1. forked the repo
2. downloaded my fork to my pc
3. made the changes and committed locally (after some time/days)
before I committing to my fork, I recognized that the main repo (this repo here)
and my fork out of sync so I did a compare of this repo on my fork,
created a PR in my fork and merged the changes from main to be in sync, 
but now my fork is telling me, that it is one commit ahead of the main repo.
Where is my error? How can I go in sync with this repo?
I searched the web and there are some hints to create another upstream repo ...
is this still true? Is the workflow
1. fork a repo
2. create a local repo from the forked one
3. create an upstream repo link
4. before creating a PR get the local repo in sync with upstream repo
5. commit changes from the upstream repo to the forked repo
6. do changes in the local repo
7. commit local changes to forked repo
Really?

**Comment by chcg** [comment]
@Ekopalypse 

4. Create a branch from your forked master to do your changes
5. ?? What do you mean with commit changes from upstream to forked? Sync your master with upstream aka pull upstream\master and push to origin\master?
 
Typically you want to do a rebase of your changes to the current upstream\master version. But that is just necessary if there are conflicts with the merging (see e.g. https://github.com/notepad-plus-plus/notepad-plus-plus/pull/3393 for such a conflict). 

So if you didn't started from a very old version first create the PR and then check if there are conflicts which need to be resolved by merging or rebasing.

**Comment by suby-ge** [comment]
https://community.notepad-plus-plus.org/topic/17256/guide-how-to-install-the-pythonscript-plugin-on-notepad-7-6-3-7-6-4-and-above

This worked for me like a charm! 👍
